lmdeploy/serve/turbomind/deploy.py

```diff
@@ -379,172 +143,892 @@ def permute(x: torch.Tensor):
+                                model_params.pop(
+                                    f'layers.{i}.attention.{fb_key}.{t}')
+                    except:  # noqa: E722
+                        pass
+
+                    if t == 'bias':
+                        continue
+                    # ffn
+                    for key in ['w1', 'w2', 'w3']:
+                        params[i][f'ffn.{key}.weight'] = \
+                            model_params.pop(
+                                f'layers.{i}.feed_forward.{key}.{t}')
+                    # norm
+                    params[i]['attn_norm.weight'] = \
+                        model_params.pop(f'layers.{i}.attention_norm.{t}')
+                    params[i]['ffn_norm.weight'] = \
+                        model_params.pop(f'layers.{i}.ffn_norm.{t}')
+
+            others = ['tok_embeddings.weight', 'norm.weight', 'output.weight']
+            for key in others:
+                params[key] = model_params.pop(key)
+            self.params = params
+
+        def layers(self):
+            for layer_id in range(self.num_layer):
+                yield LlamaDecoderLayer(layer_id, self.params[layer_id])
+
+    def bins(self):
+        for bin in self.bin_files:
+            yield bin
+
+
+class HfModel:
+
+    def __init__(self, model_path: str, tokenizer_path: str, **kwargs):
+        self.model_path = model_path
+        self.tokenizer_path = tokenizer_path
+
+        if 'quant_path' in kwargs and kwargs['quant_path'] is not None:
+            bin_files = kwargs['quant_path']
+        else:
+            bin_files = [
+                file for file in os.listdir(model_path)
+                if file.endswith('.bin')
+            ]
+        bin_files = sorted(bin_files)
+        self.bin_files = [os.path.join(model_path, file) for file in bin_files]
+        self.model_info = self.get_model_info()
+
+    def tokenizer_info(self):
+        return tokenizer_info_sp(self.tokenizer_path)
 
-    _qweight = 'weight'
-    _suffixes = [_qweight, 'bias']
+    def get_model_info(self):
+        # read model arguments from params.json
+        try:
+            params_path = osp.join(self.model_path, 'config.json')
+            with open(params_path) as f:
+                model_arg = json.load(f)
+                num_layer = model_arg['num_hidden_layers']
+                norm_eps = model_arg['rms_norm_eps']
+                if 'num_key_value_heads' in model_arg:
+                    kv_head_num = model_arg['num_key_value_heads']
+                else:
+                    kv_head_num = model_arg['num_attention_heads']
+        except Exception as e:
+            print(f'get "num_hidden_layers" and "rms_norm_eps" from '
+                  f'{params_path} failed: {e}')
+            assert (0)
+        return dict(num_layer=num_layer,
+                    norm_eps=norm_eps,
+                    kv_head_num=kv_head_num)
+
+    class BinFile:
+
+        def __init__(self, bin_path, unused_params):
```

Comment:
It's better to avoid `unused_params`. How about using `def __init__(self, bin_path, **kwargs)`?

Reply:
unused_params is unused tensor in checkpoint.